File: 3090-maximum-length-substring-with-two-occurrences/3090-maximum-length-substring-with-two-occurrences.go
func maximumLengthSubstring(s string) int {
    ans := 0
    freq := map[byte]int{}
    left := 0
    for right := 0; right < len(s); right++ {
        freq[s[right]]++

        for freq[s[right]] > 2 {
            freq[s[left]]--
            left++
        }

        ans = max(ans, right - left + 1)
    }

    return ans
}

// hash map, sliding window, string
// time: O(n)
// space: O(k)

File: 3090-maximum-length-substring-with-two-occurrences/3090-maximum-length-substring-with-two-occurrences.py
class Solution:
    def maximumLengthSubstring(self, s: str) -> int:
        ans = 0
        freq = {}
        left = 0
        for right in range(len(s)):
            if s[right] not in freq:
                freq[s[right]] = 1
            else:
                freq[s[right]] += 1
            
            while freq[s[right]] > 2:
                freq[s[left]] -= 1
                left += 1
            
            ans = max(ans, right - left + 1)
        
        return ans

# hash map, sliding window, string
# time: O(n)
# space: O(1)